staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go

```diff
@@ -107,16 +206,78 @@ func ConvertToOpenstackRequest(body []byte) ([]byte, error) {
 		return nil, err
 	}
 
-	klog.Infof("debug: request struct: %v", obj)
-
 	cpu, mem := GetRequestedResource(obj.Server.Flavor)
 	imageUrl := GetVmImageUrl(obj.Server.ImageRef)
 
 	podJson := fmt.Sprintf(POD_JSON_STRING_TEMPLATE, obj.Server.Name, imageUrl, obj.Server.Name, cpu, mem, cpu, mem)
-	klog.Infof("debug: pod json: %s", podJson)
+	klog.V(6).Infof("pod json: %s", podJson)
 	return []byte(podJson), nil
 }
 
+// Convert the action request to Arktos action request body
+func ConvertActionFromOpenstackRequest(body []byte) ([]byte, error) {
+	op := getActionOperation(body)
+	klog.V(6).Infof("Convert %s Action", op)
+
+	switch op {
+	case REBOOT:
+		obj := ArktosReboot{"v1", "CustomAction", op, ArktosRebootParams{10}}
+		return json.Marshal(obj)
+	case SNAPSHOT:
+		o := OpenstackCreateImageRequest{}
+		err := json.Unmarshal(body, &o)
+		if err != nil {
+			return nil, fmt.Errorf("invalid snapshot request. error %v", err)
+		}
+		obj := ArktosSnapshot{"v1", "CustomAction", op, ArktosSnapshotParams{o.Snapshot.Name}}
+		return json.Marshal(obj)
+	case RESTORE:
+		o := OpenstackRebuildRequest{}
+		err := json.Unmarshal(body, &o)
+		if err != nil {
+			return nil, fmt.Errorf("invalid restore request. error %v", err)
+		}
+		obj := ArktosRestore{"v1", "CustomAction", op, ArktosRestoreParams{o.Restore.ImageRef}}
+		return json.Marshal(obj)
+	default:
+		return nil, fmt.Errorf("unsupported action: %s", op)
+	}
+}
+
+func getActionOperation(body []byte) string {
+	for _, action := range supportedActions {
+		pattern := fmt.Sprintf(`"%s" *:`, action)
+		match, _ := regexp.Match(pattern, body)
+		if match {
+			return action
+		}
+	}
+
+	return "unknownAction"
+}
+
+func ConvertActionToOpenstackResponse(obj runtime.Object) runtime.Object {
+	o := obj.(*v1.Status)
+	klog.V(6).Infof("Convert Arktos object: %v", o)
+
+	// for action types reboot, start, stop, simply return empty response since Openstack
+	if strings.Contains(o.Message, REBOOT) {
+		return nil
+	}
+
+	if strings.Contains(o.Message, SNAPSHOT) {
+		s := OpenstackCreateImageResponse{ImageId: o.Details.Name}
+		return &s
+	}
+
+	if strings.Contains(o.Message, RESTORE) {
+		s := OpenstackRebuildResponse{ImageId: o.Details.Name, CreatedAt: time.Now().String()}
+		return &s
+	}
+
+	return o
```

Comment:
Perhaps log to indicate that this action does not support for now?

Reply:
the log is in the ConvertActionFromOpenstackRequest() function as the default of the action operations.